Add I2C DHT12 sensor

enelson1001 opened the pull request:
Please review adding i2c DHT12 sensor

PerMalmberg commented:
Very neat, and so little code needed for another device. 👍 

Could you please update the contributions file naming yourself and linking this PR? Also, please add this device to the readme in the project root.

Do you have more changes coming?

enelson1001 commented:
1. Will update contributors.md file.
2. Will update readme.md file
3. Will add datasheet to Documents folder.
4. I have written a driver for the SH1107 display used on the M5Stick-Mono version and I am currently working on the driver for the ST7735S that is used on the M5StickC.  I thought I would submit both of these at the same time when they were done.  On the SH1107 driver I can't see how to create a test code since the chip does not support reporting an ID when using SPI.  
4a. Will there be a problem if no test code is provided? 
4b. Is it OK to create a new PR for these two displays or do you want to add the displays to this PR?
5. I have a couple of questions for you in the next two comments (hope you don't mind)

enelson1001 commented:
Question 1
I have been running my application M5StackBME280 as seen on my github page.  In the code I print out the Systems Stats every couple of seconds.  Below is what I have captured.  My question: Is it normal for Main task to keep growing (ie use more memory)? To me it looks like the App will eventually run out of memory?  Am I correct in thinking that LittlevGL is not the problem since its stack size is not changing?

``` c
=================================================================================================
Jan 4 - using malloc and free in littlevgl
=================================================================================================
I (4755) SocketDispatcher: Network up, sockets will be restarted.
W (10773) APP: ============ Main TICK TICK TICK  =============
I (10774) MemStat: Mem type |  8-bit free | Smallest block | Minimum free | 32-bit free | Smallest block | Minimum free
I (10781) MemStat: INTERNAL |      107812 |          88200 |        97460 |      158656 |          88200 |       148296
I (10792) MemStat:      DMA |      107812 |          88200 |        97460 |      107812 |          88200 |        97460
I (10803) MemStat:   SPIRAM |           0 |              0 |            0 |           0 |              0 |            0
I (10815) MemStat: 
I (10819) MemStat:             Name |      Stack |  Min free stack |  Max used stack
I (10826) MemStat:         LvglTask |       4096 |             560 |            3536
I (10834) MemStat:         SntpTask |       3200 |             772 |            2428
I (10843) MemStat:   PollSensorTask |       3300 |             484 |            2816
I (10851) MemStat: SocketDispatcher |      20480 |           18392 |            2088
I (10859) MemStat:         MainTask |      16384 |           13308 |            3076


W (81440) APP: ============ Main TICK TICK TICK  =============
I (81441) MemStat: Mem type |  8-bit free | Smallest block | Minimum free | 32-bit free | Smallest block | Minimum free
I (81448) MemStat: INTERNAL |      107504 |          88200 |        97120 |      158348 |          88200 |       147956
I (81459) MemStat:      DMA |      107504 |          88200 |        97120 |      107504 |          88200 |        97120
I (81471) MemStat:   SPIRAM |           0 |              0 |            0 |           0 |              0 |            0
I (81482) MemStat: 
I (81485) MemStat:             Name |      Stack |  Min free stack |  Max used stack
I (81493) MemStat:         LvglTask |       4096 |             560 |            3536
I (81502) MemStat:         SntpTask |       3200 |             772 |            2428
I (81510) MemStat:   PollSensorTask |       3300 |             484 |            2816
I (81519) MemStat: SocketDispatcher |      20480 |           18216 |            2264
I (81527) MemStat:         MainTask |      16384 |           12588 |            3796

W (2040282) APP: ============ Main TICK TICK TICK  =============
I (2040283) MemStat: Mem type |  8-bit free | Smallest block | Minimum free | 32-bit free | Smallest block | Minimum free
I (2040290) MemStat: INTERNAL |      107448 |          88200 |        77764 |      158292 |          88200 |       128600
I (2040302) MemStat:      DMA |      107448 |          88200 |        77764 |      107448 |          88200 |        77764
I (2040313) MemStat:   SPIRAM |           0 |              0 |            0 |           0 |              0 |            0
I (2040325) MemStat: 
I (2040328) MemStat:             Name |      Stack |  Min free stack |  Max used stack
I (2040336) MemStat:         LvglTask |       4096 |             560 |            3536
I (2040345) MemStat:         SntpTask |       3200 |             772 |            2428
I (2040353) MemStat:   PollSensorTask |       3300 |             476 |            2824
I (2040362) MemStat: SocketDispatcher |      20480 |           18216 |            2264
I (2040371) MemStat:         MainTask |      16384 |           12524 |            3860

W (3948752) APP: ============ Main TICK TICK TICK  =============
I (3948754) MemStat: Mem type |  8-bit free | Smallest block | Minimum free | 32-bit free | Smallest block | Minimum free
I (3948761) MemStat: INTERNAL |      107332 |          88200 |        77704 |      158176 |          88200 |       128540
I (3948772) MemStat:      DMA |      107332 |          88200 |        77704 |      107332 |          88200 |        77704
I (3948783) MemStat:   SPIRAM |           0 |              0 |            0 |           0 |              0 |            0
I (3948795) MemStat: 
I (3948798) MemStat:             Name |      Stack |  Min free stack |  Max used stack
I (3948807) MemStat:         LvglTask |       4096 |             560 |            3536
I (3948816) MemStat:         SntpTask |       3200 |             772 |            2428
I (3948824) MemStat:   PollSensorTask |       3300 |             476 |            2824
I (3948832) MemStat: SocketDispatcher |      20480 |           18216 |            2264
I (3948841) MemStat:         MainTask |      16384 |           12508 |            3876

W (8684584) APP: ============ Main TICK TICK TICK  =============
I (8684585) MemStat: Mem type |  8-bit free | Smallest block | Minimum free | 32-bit free | Smallest block | Minimum free
I (8684592) MemStat: INTERNAL |      107324 |          88200 |        75800 |      158168 |          88200 |       126636
I (8684604) MemStat:      DMA |      107324 |          88200 |        75800 |      107324 |          88200 |        75800
I (8684615) MemStat:   SPIRAM |           0 |              0 |            0 |           0 |              0 |            0
I (8684627) MemStat: 
I (8684630) MemStat:             Name |      Stack |  Min free stack |  Max used stack
I (8684638) MemStat:         LvglTask |       4096 |             560 |            3536
I (8684647) MemStat:         SntpTask |       3200 |             772 |            2428
I (8684655) MemStat:   PollSensorTask |       3300 |             476 |            2824
I (8684664) MemStat: SocketDispatcher |      20480 |           18216 |            2264
I (8684673) MemStat:         MainTask |      16384 |           12508 |            3876

W (19560761) APP: ============ Main TICK TICK TICK  =============
I (19560762) MemStat: Mem type |  8-bit free | Smallest block | Minimum free | 32-bit free | Smallest block | Minimum free
I (19560770) MemStat: INTERNAL |      107120 |          88200 |        76388 |      157964 |          88200 |       127224
I (19560782) MemStat:      DMA |      107120 |          88200 |        76388 |      107120 |          88200 |        76388
I (19560793) MemStat:   SPIRAM |           0 |              0 |            0 |           0 |              0 |            0
I (19560804) MemStat: 
I (19560808) MemStat:             Name |      Stack |  Min free stack |  Max used stack
I (19560817) MemStat:         LvglTask |       4096 |             560 |            3536
I (19560825) MemStat:         SntpTask |       3200 |             772 |            2428
I (19560834) MemStat:   PollSensorTask |       3300 |             476 |            2824
I (19560842) MemStat: SocketDispatcher |      20480 |           18216 |            2264
I (19560851) MemStat:         MainTask |      16384 |           12476 |            3908
```

enelson1001 commented:
Question 2
LittlevGL requires the DisplayDriver to call lv_tick_inc(num_of_milliseconds).  In the Display driver code I have accomplished this by using the following code.
``` c
esp_register_freertos_tick_hook(lv_tick_task);  // Inside bool DisplayDriver::initialize()

and

// The  lv_tick_task that is required by LittlevGL for internal timing
void IRAM_ATTR DisplayDriver::lv_tick_task(void)
{
        lv_tick_inc(portTICK_RATE_MS);  // 1 ms tick_task
}
```
On the M5StackDHT12 App I have written but not yet pushed to my git hub, I eliminated 
_esp_register_freertos_tick_hook(lv_tick_task);_   and used the Smooth TimerExpiredEvent but this increased code size by 10Kbytes.
``` c
Inside - bool DisplayDriver::initialize()
tick = Timer::create(1, timer_expired_event, true, milliseconds{5});
tick->start();

and
void DisplayDriver::event(const smooth::core::timer::TimerExpiredEvent& event)
{
        lv_tick_inc(5);  // 5 ms tick_task
}
```
1. My question: Is it OK to use _esp_register_freertos_tick_hook(lv_tick_task);_  with Smooth without having some problems.  Right now the code runs OK with no problems but wanted to make sure no gotchas in doing so.
2. The M5StackDHT12 App using the Smooth TimerExpiredEvent also has the same problem were the MainTask keeps growing and the Internal Memory decreases.

PerMalmberg commented:
1. If memory keeps going down you have a memory leak.

2. The code increase is likely due to that you're now using the timer system. The original lv_tick_task has the IRAM_ATTR. If there's no requirement for that, then calling it via a TimerEvent isn't different from calling it from elsewhere; the TimerTick is run from within the Task that started the Tick.

enelson1001 commented:
From previous questions and your response.
1. Memory finally stabilized but took a very very long time (24 hrs).
2. My DisplayDriver is now using IRAM_ATTR lv_tick_task() and esp_register_freertos_tick_hook(lv_tick_task) and not a seperate TimerTick task.  Everything is working OK.

I thought I would give you an update on my progress.  I have successfully been able to write to the ST7735S display on the M5StickC.  I have changed the name of the ILI9341.cpp and ILI9341.h to LcdSpi.cpp and LcdSpi.h since both the ILI9341 and the ST7735 can use the same code.  The only thing that is different is the initialization commands.  I removed the back-light pin since the ST7735 on the M5StickC uses a power management IC to control back-lighting.  I do back-lighting settings in the next level up code which I call DisplayDriver.

I have written code for the I2C-AXP192 power management IC because I needed it to be able to get the M5Stick to power up properly (and power the ST7735s chip) but I need to do some more testing.

The M5Stick also has an I2C RTC - BM8563 which I plan on writing the I2C code for it.  I have not started on this yet but I don't think it will take to long to complete.

I have written separate code for the display chip SH1107 used on the M5StickMono which is working but now I would like to use the LcdSpi.  To incorporate the SH1107 into the LcdSpi I would need to add two functions: send_page() and set_common_output_scan_direction() that are only used if you are using the SH1107 and possibly other monochrome displays.  It does not use send_lines(). If I don't add these two functions to LcdSpi then I am duplicating code from the LcdSpi in the SH1107.

1. Is it OK to add  send_page() and set_common_output_scan_direction() to LcdSpi?
2. Is it OK to add all the display changes in my fork for your review now and add the additional I2C drivers to my fork later.

PerMalmberg commented:
1 Sure, why not?
2. Whichever works for you. I'm pretty busy atm so a review may take some time.

enelson1001 commented:
1. Great!!!
2. Since you are busy will dump everything into fork at the same time. May end of next week.

enelson1001 commented:
# Summary of display changes
I thought I would submit the display changes now for you review since I am still working on the AXP driver.
## 1. DisplaySpi.cpp DisplaySpi.h
This was originally called ili9341.cpp.  Re-named as this is the core class to use when interfacing to a display.  I was going to name it LcdSpi but since there are LCD and OLED displays I decided to make the name more generic.  Added preset pin levels for chip select and reset.
## 1.1 Function - set_madctl()
This function was originally named set_rotation.  Re-named to set_madctl because that is what the function is doing.  Added a argument to allow user to change the madct register address from the default of 0x36.
## 1.2 Function - hw_reset()
Added addition arguments active_low, active_time and delay so the function is allows a different active reset state and allows users to fine tune the reset times.
## 1.3 Function - sw_reset()
Added function to support software reset of display.  Some display do not use hardware reset and only have software reset.  This function could be called before the init commands are executed.
## 1.4 Function - send_init_cmds()
This function originally didn't have any arguments.  Now we pass in the init_cmds starting address and the length or number of init_cmds so user can specify.  Also cleaned up code and added the ability to have a delay when a command has data.
## 1.5 Function - send_cmds()
This is a new function added.  It was added to support the monochrome display.  This function is more efficient the doing multiple send_cmd.
## 1.6 Function - send_cmd()
This function was originally named lcd_wr_cmd.  Wanted to get away from having "lcd" in the name. The name "send" was being used so decided all write functions would have the name send.
## 1.7 Function - send_data()
This function was originally name lcd_wr_data.
## 1.8 Function - set_back_light()
This function was removed and performed by the next level up from DisplaySpi.  Some display do not control back-lighting via a GPIO pin.  The M5StickC using the AXP192 power management IC to set the output voltage on an LDO to control the back-lighting. Some displays may use a PWM signal.
## 2.1 DisplayTypes.h
A header file that contains custom typedef used on DisplaySpi.  
## 3.1 DisplayCommands.h
A header file that contain the most common LCD commands. New commands required for different displays should be included in their "MyNewDisplay.h" file.
## 4.1 IlI9341.h
A header file that contains information pertaining to the ILI9341.  In this case it only contains vector<display_int_cmd_t> of initialization commands.
## 5.1 ST7735.h
A header file that contains some unique init_cmds for various ST7735 models and some unique CGRAM offsets for various ST7735 models.
## 6.1 SH1107.h
A header file that contains unique commands for the SH1107, the init_cmds.
## 7.1 Smooth/test/spi_4_line_devices_test.cpp .h
Modified to use DisplaySpi instead of ILI9341 and removed enabling back-light.

# Reset discussion
Some displays do not use a reset pin and tie the pin HIGH.  This saves them on not have to use a gpio pin and instead use sw_reset.  I could not find a way to pass GPIO_NUM_NC for reset. Is there a way to pass GPIO_NUM_NC or should the reset function be removed from DisplaySPI like back-lighting?

# My progress
The AXP driver is taking a lot more time then I anticipated, partially because I am doing it for AXP192, AXP173 and AXP202 chips.  There are a lot of registers and lot of functionality in each register.  The code will be large so I will eventually submit but I would understand if you did not want to add an I2C driver that is well over 1500 lines.
I have not started on the I2C RTC chip.

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
-              cs_pin(chip_select_pin, true, false, false)
+              reset_pin(reset_pin, true, false, false),     // GPIO_MODE_OUTPUT, no pullup, no pulldown
+              dc_pin(data_command_pin, true, false, false), // GPIO_MODE_OUTPUT, no pullup, no pulldown
+              cs_pin(chip_select_pin, true, false, false)   // GPIO_MODE_OUTPUT, no pullup, no pulldown
```
Is this hardware dependent, i.e. does it makes sense to have these settings passed in to the class in the constructor?

enelson1001 replied:
All the display chips I have looked at require chip select (cs), data/command (dc) and possibly reset.  The constructor currently takes the gpio_num_t for these three pins.  I guess I am not sure what you are asking or looking for?

PerMalmberg replied:
My bad, ignore :)

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -26,26 +26,25 @@ using namespace smooth::core::io::spi;
+                           uint32_t spi_device_flags,
+                           int spi_queue_size,
+                           bool use_pre_transaction_callback,
+                           bool use_post_transaction_callback)
```
An awful lot of parameters. When they get this many I prefer to create a configuration class that is passed to the constructor, esp. if most of them just are copied into the class in which case I just hold a copy of the configuration class instead. Not sure if it makes things easier here though.

enelson1001 replied:
Almost all of these parameters get past to the base class SPIDevice.  Should we be looking at refractoring SPIDevice?

PerMalmberg replied:
Do you think it'll make it more readable? If yes, then that's an option.

enelson1001 replied:
From my point of view I do not mind all the parameters and to me seeing everything that is required/needed to create the SPI device one place is the best approach.
``` c 
auto device = spi_master.create_device<DisplaySpi>(
                        GPIO_NUM_14,            // chip select gpio pin
                        GPIO_NUM_27,            // data command gpio pin
                        GPIO_NUM_33,            // reset gpio pin
                        0,                      // spi command_bits
                        0,                      // spi address_bits,
                        0,                      // bits_between_address_and_data_phase,
                        0,                      // spi_mode = 0,
                        128,                    // spi positive_duty_cycle,
                        0,                      // spi cs_ena_posttrans,
                        SPI_MASTER_FREQ_16M,    // spi-sck = 16MHz
                        0,                      // full duplex (4-wire)
                        7,                      // queue_size,
                        true,                   // use pre-trans callback
                        true);                  // use post-trans callback
```
I guess we could create a struct or class to hold the spi_config parameters but I don't feel it would make the code easier to read.

PerMalmberg replied:
Then leave it.

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
-    // Reset the display
-    void ILI9341::reset_display()
+    // Hardware Reset
+    void DisplaySpi::hw_reset(bool active_low, int active_time, int delay_time)
```
Pass in `std::chrono::milliseconds` instead of an int for the time-parameters, it makes the API more clear.

enelson1001 replied:
Done

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
+    }
+
+    // Software Reset
+    bool DisplaySpi::sw_reset(int delay_time)
```
Pass in std::chrono::milliseconds as per above comment.

enelson1001 replied:
Done

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
+                size_t delay_time = init_cmds[index].data[delay_data_length];
+
+                // check delay_data_length to see if there is data to send before delaying
+                if (delay_data_length != 0)
```
Using != isn't wrong, but `delay_data_length > 0` makes more sense in this context. Change it or leave it as-is.

enelson1001 replied:
Done

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
 
     // Send initialize sequence of commands and data to display
-    bool ILI9341::send_init_cmds()
+    bool DisplaySpi::send_init_cmds(const display_init_cmd_t* init_cmds, size_t length)
```
There are several `res &= ...` in this function, but no checks if `res` becomes false. Does it make sense to perform additional commands once a function has failed or should the conditions have an added `res`-check, such as `for (size_t index = 0; res && index < length; index++)` ?

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
+
+        std::lock_guard<std::mutex> lock(get_guard());
+        spi_transaction_t trans;
+        std::memset(&trans, 0, sizeof(trans));  //Zero out the transaction
```
Use zero-intialization instead of the memset: `spi_transaction_t trans{};`

https://en.cppreference.com/w/cpp/language/zero_initialization

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
+        spi_transaction_t trans;
+        std::memset(&trans, 0, sizeof(trans));  //Zero out the transaction
+        trans.rx_buffer = nullptr;
+        trans.rxlength = 0;
```
No need to zero these, they are already zeroed out by the memset or the zero-initialization once you've changed it.

enelson1001 replied:
Done, look at new code

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -154,17 +212,17 @@ namespace smooth::application::display
-    bool ILI9341::lcd_wr_data(const uint8_t* data, size_t length)
+    // Send data to the displayc.
+    // Uses spi_device_polling_write, which waits until the transfer is complete.
+    bool DisplaySpi::send_data(const uint8_t* data, size_t length)
```
Do you have plans to make this API async, i.e. non-blocking?

enelson1001 replied:
Removed comment

PerMalmberg replied:
Nothing wrong with the comment, I was just wondering :)

PerMalmberg commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
-        cs_posttrans_pin_states.at(0) = PIN_HIGH;
+        dc_pretrans_pin_states[0] = PIN_LOW;
+        cs_pretrans_pin_states[0] = PIN_LOW;
+        cs_posttrans_pin_states[0] = PIN_HIGH;
```
These three lines are the same as in `send_cmds` (very similar named btw) and there's a nearly identical set in `send_data` I think it makes sense to break them into a `prepare_pin_command()` and `prepare_pin_data()` or similarly named functions, thoughts?

enelson1001 replied:
Created two new functions,

PerMalmberg commented on lib/smooth/include/smooth/application/display/DisplayTypes.h:
```diff
@@ -0,0 +1,38 @@
+    {
+        uint16_t col_offset;
+        uint16_t row_offset;
+    } cgram_offset_t;
```
I don't think you need the typedef keyword in C++?

https://en.cppreference.com/w/c/language/struct

enelson1001 replied:
Removed typedef replaced with using ...

PerMalmberg commented on lib/smooth/include/smooth/application/display/ST7735.h:
```diff
@@ -0,0 +1,147 @@
+    ************************************************************************************/
+
+    // init commands for black tab - 17 commands
+    static const std::vector<display_init_cmd_t> init_cmds_7735B =
```
If you make these into `static constexpr std::array<...>` you can save memory at runtime since the compiler will be able to determine which parts of the array you actually use.

enelson1001 replied:
Done

PerMalmberg commented:
I left a few comments, but overall it looks really good.

> Some displays do not use a reset pin and tie the pin HIGH. This saves them on not have to use a gpio pin and instead use sw_reset. I could not find a way to pass GPIO_NUM_NC for reset. Is there a way to pass GPIO_NUM_NC or should the reset function be removed from DisplaySPI like back-lighting?

I don't understand what it is you want to do? If they tie a pin high, they're not saving a GPIO, are they? Or do you mean they tie it high on the display side, that'd save a pin on the MCU.

enelson1001 commented:
Yes this is what I mean - they tie the reset pin high on the display side and that saves a pin on the MCU side.
I am kinda leaning on removing the reset pin from DisplaySPI and therefore removing the hw_reset function().

PerMalmberg commented:
From an API standpoint it's nice to provide the reset-functionality. Optimally it should be agnostic, i.e. don't care what the actual implementation is, i.e. h/w or s/w. Another way to implement it is to template the driver such that you do DispalySPI<HWreset> or DisplaySPI<SWReset>. This would make the class oblivious to the actual implementation of the reset functionality.

Don't you ever call the reset from within the class?

PerMalmberg commented:
Have you pushed the latest changes, I can't see them in this PR?

enelson1001 commented:
Yes I have pushed and see the changes in my fork.

enelson1001 commented:
You said
*From an API standpoint it's nice to provide the reset-functionality. Optimally it should be agnostic, i.e. don't care what the actual implementation is, i.e. h/w or s/w. Another way to implement it is to template the driver such that you do DispalySPI or DisplaySPI. This would make the class oblivious to the actual implementation of the reset functionality.*

*Don't you ever call the reset from within the class?*
1. I am not sure how to use templates that would allow the user to use only 2 smooth::core:io:Ouput (cs, dc) and not use (reset) smooth::core::io::Output.
2. I never call reset (sw_reset or hw_reset) within the class that is up to the user to decide which one to use.
3. An example of a display that would not require a gpio pin for reset. In the schematic you can see they use the AXP803 to perform the resetting of the tft display.  https://learn.adafruit.com/adafruit-2-4-tft-touch-screen-featherwing/downloads

PerMalmberg commented:
Odd, I see the changes in your master, but not in this PR which references your master.

Anyway, changes looks good. Re. reset, do what you feel is best.

enelson1001 commented:
The new code uploaded allows the user to add a reset pin and a a backlight pin. I also put back in the set_back_light().  The spi_4_line_devices_test.cpp shows how to add for example the reset pin.  If you do not like this approach I can revert back to the previous code or modify if you have some changes.

PerMalmberg commented:
The approach itself isn't bad, though for the sake of keeping memory usage low, I suggest using a fixed array with constants to index into the array to get the desired pin instead of the dynamically allocating hash-table.

enelson1001 commented:
Added AxpPMU as a generic way to handle the AXP173, AXP192 and AXP202.  I didn't like the idea of having 3000 lines of code to read and write every little detail on all the registers on the AXP.  Almost all the registers are used to configuring the chip so took the approach that the lcd displays used and created a function - write_int_regs() that would be used to write a bunch of registers to configure the chip.  The user will create this file since it is product dependent meaning I can't create  for example an AXP192InitCmd.h file because one product may program the LDO or DC-DC power supplies to a different voltage. I made some functions virtual so user could inherit the AxpPMU and override these functions (some device does not support the get_gpioX_voltage) and possibly log an error.
I also added more pdf files for displays and AXP devices.  I will start on the RTC next.

PerMalmberg commented on lib/smooth/application/io/i2c/AxpPMU.cpp:
```diff
@@ -0,0 +1,350 @@
+    }
+
+    // Class constants
+    //static const char* TAG = "AxpPMU";
```
Remove if not used.

enelson1001 replied:
OK will remove

PerMalmberg commented on lib/smooth/application/io/i2c/AxpPMU.cpp:
```diff
@@ -0,0 +1,350 @@
+        uint32_t dischrg_coulomb;
+        uint8_t adc_sample_rate;
+
+        //float rate;
```
remove

enelson1001 replied:
OK will remove

PerMalmberg commented on lib/smooth/include/smooth/application/display/DisplaySpi.h:
```diff
@@ -166,14 +165,14 @@ namespace smooth::application::display
             smooth::core::io::Output dc_pin;
             smooth::core::io::Output cs_pin;
+            std::unique_ptr<DisplayPin> backlight_pin{ nullptr };
+            std::unique_ptr<DisplayPin> reset_pin{ nullptr };
```
Zero-initialize: `backlight{}`, no need to specify nullptr.

enelson1001 replied:
Ok will remove nullptr to both

PerMalmberg commented on lib/smooth/include/smooth/application/io/i2c/AxpRegisters.h:
```diff
@@ -0,0 +1,142 @@
+        uint8_t data;
+    };
+
+    using axp_init_reg_t = AxpInitRegType;
```
Aliasing the struct seems unnecessary to me since it is just a name change, i.e. it doesn't reduce clutter. How about using `AxpInitReg` throughout the class?

enelson1001 replied:
OK

PerMalmberg commented on lib/smooth/include/smooth/application/io/i2c/AxpRegisters.h:
```diff
@@ -0,0 +1,142 @@
+
+namespace smooth::application::sensor
+{
+    enum class Register : uint8_t
```
`Register` is a very generic name, how about naming it `AxpRegister` to avoid future name clashes? Or, place the entire Axp code inside its own namespace?

enelson1001 replied:
OK renamed to AxpRegister

enelson1001 commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -93,22 +93,20 @@ namespace smooth::application::display
-        auto search = display_pins.find("reset");
-
-        if (search != display_pins.end())
+        if (reset_pin != nullptr)
```
The code in my repository already had this change, Strange

enelson1001 commented on lib/smooth/application/display/DisplaySpi.cpp:
```diff
@@ -424,23 +422,21 @@ namespace smooth::application::display
-        auto search = display_pins.find("bklight");
-
-        if (search != display_pins.end())
+        if (backlight_pin != nullptr)
```
The code in my repository already had this change, Strange.

enelson1001 commented:
I will create README-AXP.md and submit updates soon

enelson1001 commented:
I won't be added anymore devices (maybe a couple more but that will be a few months from now) and think I have made all the changes you requested for AxpPMU and DisplaySPI.  I wasn't sure where to place the README-AXP file so I just put in Documents.  I assume one more round of changes for the RTC and hopefully you can push to your branch so I can update my projects.   I have to say I am impressed with how much functionality I can get using Smooth and LvGL on a ESP32 with and LCD display.

PerMalmberg commented on test/i2c_rtc8563_test/i2c_rtc8563_test.cpp:
```diff
@@ -0,0 +1,222 @@
+        }
+    }
+
+    // Set alarm - alarm will activate 12 minutes past any hour of any day of any month of any weekday.
```
This comment doesn't match with the values set below.

enelson1001 replied:
If you are referring to this comment
```
// Set alarm - alarm will activate 12 minutes past any hour of any day of any month of any weekday.
```
The comment is correct since I only enable the alarm_minute.
```
tm.ena_alrm_minute = true;
tm.ena_alrm_hour = false;
tm.ena_alrm_day = false;
tm.ena_alrm_weekday = false;
```

PerMalmberg replied:
Oh, ok. A bit confusing, with the other values being set, but ok.

PerMalmberg commented:
>  I assume one more round of changes for the RTC

Just a small comment, looks fine otherwise.

> I have to say I am impressed with how much functionality I can get using Smooth and LvGL on a ESP32 with and LCD display.

Me too, I never thought it'd be so useful for anyone else. I'm a bit surprised that you've not found any bugs, considering that entire Smooth (well, almost now) has been a one-man project. Sure, I've put in a lot of effort into it but still... It makes me happy to see it being used.

PerMalmberg commented:
Merged and done. Awesome work, @enelson1001
